stan/math/rev/fun/tcrossprod.hpp

```diff
@@ -27,35 +27,20 @@ tcrossprod(const T& M) {
   if (M.rows() == 0) {
     return {};
   }
-  // if (M.rows() == 1)
-  //   return M * M.transpose();
-
-  // WAS JUST THIS
-  // matrix_v result(M.rows(), M.rows());
-  // return result.setZero().selfadjointView<Eigen::Upper>().rankUpdate(M);
-
-  Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime> MMt(M.rows(),
-                                                                     M.rows());
-
-  vari** vs
-      = reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
-          (M.rows() * M.cols()) * sizeof(vari*)));
-  int pos = 0;
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < M.cols(); ++n) {
-      vs[pos++] = M(m, n).vi_;
-    }
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    MMt.coeffRef(m, m)
-        = var(new internal::dot_self_vari(vs + m * M.cols(), M.cols()));
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < m; ++n) {
-      MMt.coeffRef(n, m) = MMt.coeffRef(m, n) = dot_product(M.row(m), M.row(n));
-    }
-  }
-  return MMt;
+
+  arena_matrix<promote_scalar_t<var, T>> arena_M = M;
+  arena_matrix<promote_scalar_t<double, T>> arena_M_val = value_of(arena_M);
+
+  Eigen::MatrixXd res_val(M.rows(), M.rows());
+  arena_matrix<promote_scalar_t<var, T>> res
+      = arena_M_val * arena_M_val.transpose();
+
+  reverse_pass_callback([res, arena_M, arena_M_val]() mutable {
+    ref_type_t<decltype(res.adj())> adj = res.adj();
```

Comment:
How should we be future-proofing these things for `var_value<MatrixXd>` (when getting the adjoint out of a var/matrix thing is easy)? Does this look right?

Reply:
I would only do that if `adj` is used more than once. Otherwise leaving it an expression by making its type `const auto&` or even not making it a separate variable is preferable.

Also, when you do want a ref this is simpler: `const auto& adj = to_ref(res.adj());`

Reply:
For `var_value<MatrixXd>` the `ref_type_t<decltype(res.adj())>` should just be the same as `adj_`'s type (`Eigen::Map<double, R, C>`) right? If so idt this would need any changes